### PR TITLE
PyTorch Execution Graph Observers

### PR DESCRIPTION
**Pull request by louisfeng**
Summary:
This diff adds an execution graph observer that tracks all operators (dispatcher autograd, jit, user defined, etc.) and their inputs and outputs. The results are written to a temp JSON file which can be used for further analysis. This support various use cases, such as dependency analysis, performance optimizations, etc.

Some minor refactoring of existing code for clarity and completeness.

Test Plan:
Example output:

{F580073970}

```
=> buck build caffe2/torch/fb/observers:execution_graph_observer_runner --show-output

=> buck-out/gen/caffe2/torch/fb/observers/execution_graph_observer_runner --pytorch_enable_execution_graph_observer=true --pytorch_execution_graph_observer_iter_label="## START ##" --pytorch_execution_graph_observer_iter_target=3
I0412 23:01:34.426760 2495964 ExecutionGraphObserver.cpp:391] Enabled PyTorch execution graph observer
I0412 23:01:34.427108 2495964 ExecutionGraphObserver.cpp:394] Matching iteration start label: "## START ##"
I0412 23:01:34.427227 2495964 ExecutionGraphObserver.cpp:405] Target iteration: 3
I0412 23:01:34.427540 2495964 ExecutionGraphObserverRunner.cpp:50] Running test execution graph observer runner.
I0412 23:01:34.427655 2495964 ExecutionGraphObserverRunner.cpp:51] iterations: 10
I0412 23:01:34.771853 2495964 ExecutionGraphObserver.cpp:134] Writing PyTorch execution graph to: /tmp/pytorch_execution_graph_2495964_3.json
I0412 23:01:34.811157 2495964 ExecutionGraphObserver.cpp:304] PyTorch execution graph is written to file: /tmp/pytorch_execution_graph_2495964_3.json
```

see `/tmp/pytorch_execution_graph_[process_id]_[iter_target].json`

Differential Revision: D27238906

### CODE REVIEW

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -119,11 +119,9 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
...
     }
 
-    if (profiler::profilerEnabled()) {
-      // If profiler is enabled, thread_id is stored.
-      // See NOTE [ Sequence Numbers ]
-      thread_id_ = at::RecordFunction::currentThreadId();
-    }
+    // Store the thread_id of the forward operator.
+    // See NOTE [ Sequence Numbers ]
+    thread_id_ = at::RecordFunction::currentThreadId();
```
Comment (albanD) [inline torch/csrc/autograd/function.h:124]: Why do you need to change that?
What is the perf cost of always doing this lookup?

**Review by albanD — CHANGES_REQUESTED** [review]
This function is very perf sensitive, we should have a benchmark that ensures there is no bad regression due to this change before landing.

File: torch/csrc/autograd/function.h (hunk 1 above)
Comment (louisfeng) [reply torch/csrc/autograd/function.h:124]: Hi @albanD, `profilerEnabled()` and profiler_legacy is being deprecated (please see this [conversation](https://www.internalfb.com/diff/D27238906?dst_version_fbid=132461845501449&transaction_fbid=445031943390803)). `currentThreadId()` returns the an uint64_t, and is set once per thread. The cost should be almost the same as we are removing a conditional and `currentThreadId()` internally does a check on the uint64_t has been set and returns the value.

**Comment by louisfeng** [comment]
> This function is very perf sensitive, we should have a benchmark that ensures there is no bad regression due to this change before landing.

Please see my comment above.

File: torch/csrc/autograd/function.h (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/function.h:124]: I understand you want to remove that function, but we should still make sure there isn't any perf issue related to this.
You can use: https://pytorch.org/docs/master/benchmark_utils.html?highlight=timer#torch.utils.benchmark.Timer in instruction count mode to easily benchmark this.

Comment (louisfeng) [reply torch/csrc/autograd/function.h:124]: Sure, I could run a comparison for before/after. Do you have any recommendation on which benchmarks use for this?

Comment (albanD) [reply torch/csrc/autograd/function.h:124]: Any code that operates on Tensors that require gradients will end up here, so something like this should work:
```
a = torch.rand(1, requires_grad=True)

b = a + 1
```

Comment (louisfeng) [reply torch/csrc/autograd/function.h:124]: Thanks, I just copied the examples from the benchmark tutorial page and added `requires_grad=True`. 

Source code and instruction count comparison:

P404935604

To summarize:
```
### BEFORE
                           All          Noisy symbols removed
    Instructions:      3107840                    3085414
    Baseline:             4252                       3844

### AFTER
                           All          Noisy symbols removed
    Instructions:      3101781                    3075609
    Baseline:             4236                       3844
```
So it looks like we get some savings in terms of instructions.

Comment (albanD) [reply torch/csrc/autograd/function.h:124]: Awesome, thanks for checking!

**Review by albanD — APPROVED** [review]
The PyTorch change lgtm!
